### PR TITLE
zfs: Improve import handling

### PR DESCRIPTION
**Pull request by Baughn**
###### Motivation for this change

With root on a fast device (SSD/NVMe), the import scripts start running well before the HDDs they need have been seen by the kernel. Empirically, that takes about ten seconds.

This patch increases the timeout on the root mount to 1 minute, and adds a 1-minute loop for non-root imports.

It's tested (and developed) on stable, and should be back-ported.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by Mic92** [comment]
I wish we could leverage modern approaches such as udev instead of these hacky shell scripts...
But it seems zfsonlinux still has no implementation for that.
cc @woffs

File: nixos/modules/tasks/filesystems/zfs.nix [hunk 1]
```diff
@@ -393,7 +393,13 @@ in
             };
             script = ''
               zpool_cmd="${packages.zfsUser}/sbin/zpool"
-              ("$zpool_cmd" list "${pool}" >/dev/null) || "$zpool_cmd" import -d ${cfgZfs.devNodes} -N ${optionalString cfgZfs.forceImportAll "-f"} "${pool}"
+              echo -n "importing ZFS pool \"${pool}\"..."
+              "$zpool_cmd" list "${pool}" >/dev/null 2>/dev/null && exit
+              for trial in `seq 1 60`; do
```
Comment (Mic92) [inline nixos/modules/tasks/filesystems/zfs.nix]: There should be a comment here, why this is a loop.

File: nixos/modules/tasks/filesystems/zfs.nix [hunk 2]
```diff
@@ -393,7 +393,13 @@ in
...
+                "$zpool_cmd" import -d ${cfgZfs.devNodes} -N ${optionalString cfgZfs.forceImportAll "-f"} "${pool}" && break
+                sleep 1
+              done
+              echo "Successfully imported ${pool}"
```
Comment (woffs) [inline nixos/modules/tasks/filesystems/zfs.nix]: It will print "Successfully" even if the import fails :)

Comment (Baughn) [reply nixos/modules/tasks/filesystems/zfs.nix]: Fixed.

File: nixos/modules/tasks/filesystems/zfs.nix (hunk 1 above)
Comment (Baughn) [reply nixos/modules/tasks/filesystems/zfs.nix]: Fixed.

**Comment by Baughn** [comment]
> Does "Started udev Wait for Complete Device Initialization." appear before or after "Starting Import ZFS pool" in your system log?

Before, almost instantly after kernel initialization. That just runs `udevadm settle`, which waits for the kernel and udev to finish setting up all devices that the system knows about *at that time*.

It doesn't work for HDDs, since many HDDs won't even try connecting to the bus until they've finished spinning up and initializing themselves. It's that process which takes ten seconds, and there's no way to know up-front how many such devices we're waiting for.

The only way to improve on this would be to bake in a list of HDDs we should wait for, based on the zpool output at the time we build these scripts, and I'm hesitant to call that an improvement. For one thing, it'd fail of one of those HDDs has failed. The approach I implemented, though brute-force in a sense, should also count as robust.

**Comment by Baughn** [comment]
The system log, with this patch, looks somewhat like this:
```
Jun 16 02:37:00 tsugumi systemd[1]: Started udev Wait for Complete Device Initialization.
Jun 16 02:37:00 tsugumi audit[1]: SERVICE_START pid=1 uid=0 auid=4294967295 ses=4294967295 msg='unit=systemd-udev-settle comm="systemd" exe="/nix/store/6dkz6azki0bag13m6vccy7bi41n9dak3-systemd-237/lib/>
Jun 16 02:37:00 tsugumi kernel: kauditd_printk_skb: 244 callbacks suppressed
Jun 16 02:37:00 tsugumi kernel: audit: type=1130 audit(1529113020.589:95): pid=1 uid=0 auid=4294967295 ses=4294967295 msg='unit=systemd-udev-settle comm="systemd" exe="/nix/store/6dkz6azki0bag13m6vccy7>
Jun 16 02:37:00 tsugumi systemd[1]: Starting Import ZFS pool "stash"...
Jun 16 02:37:00 tsugumi zfs-import-stash-start[2307]: importing ZFS pool "stash"...cannot import 'stash': no such pool available
Jun 16 02:37:01 tsugumi zfs-import-stash-start[2307]: cannot import 'stash': no such pool available
Jun 16 02:37:02 tsugumi zfs-import-stash-start[2307]: cannot import 'stash': no such pool available
Jun 16 02:37:03 tsugumi zfs-import-stash-start[2307]: cannot import 'stash': no such pool available
Jun 16 02:37:03 tsugumi kernel: ata1: SATA link up 3.0 Gbps (SStatus 123 SControl 300)
Jun 16 02:37:03 tsugumi kernel: ata1.00: failed to IDENTIFY (I/O error, err_mask=0x100)
Jun 16 02:37:04 tsugumi zfs-import-stash-start[2307]: cannot import 'stash': no such pool available
Jun 16 02:37:05 tsugumi zfs-import-stash-start[2307]: cannot import 'stash': no such pool available
Jun 16 02:37:06 tsugumi kernel: random: crng init done
Jun 16 02:37:06 tsugumi kernel: random: 7 urandom warning(s) missed due to ratelimiting
Jun 16 02:37:06 tsugumi zfs-import-stash-start[2307]: cannot import 'stash': no such pool available
Jun 16 02:37:07 tsugumi zfs-import-stash-start[2307]: cannot import 'stash': no such pool available
Jun 16 02:37:08 tsugumi zfs-import-stash-start[2307]: cannot import 'stash': no such pool available
Jun 16 02:37:09 tsugumi zfs-import-stash-start[2307]: cannot import 'stash': no such pool available
Jun 16 02:37:09 tsugumi kernel: ata1: SATA link up 3.0 Gbps (SStatus 123 SControl 300)
Jun 16 02:37:09 tsugumi kernel: ata1.00: ATA-7: INTEL SSDSA2M040G2GC, 2CV102HD, max UDMA/133
Jun 16 02:37:09 tsugumi kernel: ata1.00: 78165360 sectors, multi 16: LBA48 NCQ (depth 31/32)
Jun 16 02:37:09 tsugumi kernel: ata1.00: configured for UDMA/133
Jun 16 02:37:09 tsugumi kernel: scsi 0:0:0:0: Direct-Access     ATA      INTEL SSDSA2M040 02HD PQ: 0 ANSI: 5
Jun 16 02:37:09 tsugumi kernel: ata1.00: Enabling discard_zeroes_data
Jun 16 02:37:09 tsugumi kernel: sd 0:0:0:0: [sda] 78165360 512-byte logical blocks: (40.0 GB/37.3 GiB)
Jun 16 02:37:09 tsugumi kernel: sd 0:0:0:0: [sda] Write Protect is off
Jun 16 02:37:09 tsugumi kernel: sd 0:0:0:0: [sda] Mode Sense: 00 3a 00 00
Jun 16 02:37:09 tsugumi kernel: sd 0:0:0:0: [sda] Write cache: enabled, read cache: enabled, doesn't support DPO or FUA
Jun 16 02:37:09 tsugumi kernel: ata1.00: Enabling discard_zeroes_data
Jun 16 02:37:09 tsugumi kernel:  sda: sda1 sda2
Jun 16 02:37:09 tsugumi kernel: ata1.00: Enabling discard_zeroes_data
Jun 16 02:37:09 tsugumi kernel: sd 0:0:0:0: [sda] Attached SCSI disk
Jun 16 02:37:10 tsugumi kernel: ata2: SATA link up 6.0 Gbps (SStatus 133 SControl 300)
Jun 16 02:37:10 tsugumi kernel: ata2.00: ATA-10: ST8000DM002-1YW112, DN02, max UDMA/133
Jun 16 02:37:10 tsugumi kernel: ata2.00: 15628053168 sectors, multi 16: LBA48 NCQ (depth 31/32), AA
Jun 16 02:37:10 tsugumi kernel: ata2.00: configured for UDMA/133
Jun 16 02:37:10 tsugumi kernel: scsi 1:0:0:0: Direct-Access     ATA      ST8000DM002-1YW1 DN02 PQ: 0 ANSI: 5
Jun 16 02:37:10 tsugumi kernel: sd 1:0:0:0: [sdb] 15628053168 512-byte logical blocks: (8.00 TB/7.28 TiB)
Jun 16 02:37:10 tsugumi kernel: sd 1:0:0:0: [sdb] 4096-byte physical blocks
Jun 16 02:37:10 tsugumi kernel: sd 1:0:0:0: [sdb] Write Protect is off
Jun 16 02:37:10 tsugumi kernel: sd 1:0:0:0: [sdb] Mode Sense: 00 3a 00 00
Jun 16 02:37:10 tsugumi kernel: sd 1:0:0:0: [sdb] Write cache: enabled, read cache: enabled, doesn't support DPO or FUA
Jun 16 02:37:10 tsugumi kernel:  sdb: sdb1 sdb9
Jun 16 02:37:10 tsugumi kernel: sd 1:0:0:0: [sdb] Attached SCSI disk
Jun 16 02:37:10 tsugumi kernel: ata3: SATA link up 6.0 Gbps (SStatus 133 SControl 300)
Jun 16 02:37:10 tsugumi kernel: ata3.00: ATA-9: WDC WD80EFZX-68UW8N0, 83.H0A83, max UDMA/133
Jun 16 02:37:10 tsugumi kernel: ata3.00: 15628053168 sectors, multi 16: LBA48 NCQ (depth 31/32), AA
Jun 16 02:37:10 tsugumi kernel: ata3.00: configured for UDMA/133
Jun 16 02:37:10 tsugumi kernel: scsi 2:0:0:0: Direct-Access     ATA      WDC WD80EFZX-68U 0A83 PQ: 0 ANSI: 5
Jun 16 02:37:10 tsugumi kernel: sd 2:0:0:0: [sdc] 15628053168 512-byte logical blocks: (8.00 TB/7.28 TiB)
Jun 16 02:37:10 tsugumi kernel: sd 2:0:0:0: [sdc] 4096-byte physical blocks
Jun 16 02:37:10 tsugumi kernel: sd 2:0:0:0: [sdc] Write Protect is off
Jun 16 02:37:10 tsugumi kernel: sd 2:0:0:0: [sdc] Mode Sense: 00 3a 00 00
Jun 16 02:37:10 tsugumi kernel: sd 2:0:0:0: [sdc] Write cache: enabled, read cache: enabled, doesn't support DPO or FUA
Jun 16 02:37:10 tsugumi kernel:  sdc: sdc1 sdc9
Jun 16 02:37:10 tsugumi kernel: sd 2:0:0:0: [sdc] Attached SCSI disk
Jun 16 02:37:11 tsugumi kernel: ata4: SATA link up 6.0 Gbps (SStatus 133 SControl 300)
Jun 16 02:37:11 tsugumi kernel: ata4.00: ATA-9: WDC WD80EFZX-68UW8N0, 83.H0A83, max UDMA/133
Jun 16 02:37:11 tsugumi kernel: ata4.00: 15628053168 sectors, multi 16: LBA48 NCQ (depth 31/32), AA
Jun 16 02:37:11 tsugumi kernel: ata4.00: configured for UDMA/133
Jun 16 02:37:11 tsugumi kernel: scsi 3:0:0:0: Direct-Access     ATA      WDC WD80EFZX-68U 0A83 PQ: 0 ANSI: 5
Jun 16 02:37:11 tsugumi kernel: sd 3:0:0:0: [sdd] 15628053168 512-byte logical blocks: (8.00 TB/7.28 TiB)
Jun 16 02:37:11 tsugumi kernel: sd 3:0:0:0: [sdd] 4096-byte physical blocks
Jun 16 02:37:11 tsugumi kernel: sd 3:0:0:0: [sdd] Write Protect is off
Jun 16 02:37:11 tsugumi kernel: sd 3:0:0:0: [sdd] Mode Sense: 00 3a 00 00
Jun 16 02:37:11 tsugumi kernel: sd 3:0:0:0: [sdd] Write cache: enabled, read cache: enabled, doesn't support DPO or FUA
Jun 16 02:37:11 tsugumi kernel:  sdd: sdd1 sdd2
Jun 16 02:37:11 tsugumi kernel: sd 3:0:0:0: [sdd] Attached SCSI disk
Jun 16 02:37:11 tsugumi kernel: ata5: SATA link down (SStatus 0 SControl 300)
```

**Comment by woffs** [comment]
Thanks for your fixes. I think, it is ok to retry the import this way, if we don't want to insert a configurable sleep time or something similar.

**Comment by Baughn** [comment]
@Mic92 poke. I believe the review is complete?

**Comment by Baughn** [comment]
Maybe not. One failure mode that just occurred to me:

If a redundant device is missing, then `zpool import` might succeed with the pool in a degraded state, possibly with zero redundancy. I asked in #zfs, and they don't believe it'll automatically fix itself, so we need code for that case as well.

Using the udev dependencies would be a possible workaround, but I can't find documentation on that, plus we'd have the opposite problem of failing to boot if a single disk in a redundant pool has failed. Maybe. At least I can test that; it might time out and proceed to boot anyway.

Please don't accept this PR yet; I want to do some testing. I'll get back to you in <1 day.

**Comment by Baughn** [comment]
After more thought, I think the correct course of action is:

- Setting `scsi_mod.scan=sync` on the kernel command line, for both stable and unstable. I'll test if this works, and supply a patch later today if it does.

- Longer-term, using systemd device units to wait for the precise devices needed. The "longer" term will likely just be a day or two, but I don't think we should backport *this* patch, as it implies an interface change. We can use `lsblk -f` to get the exact devices needed, but this can't be run at build time -- the build might be happening on a different computer. So it has to go in nixos-generate-config.

Closing this PR, for now.

**Comment by Baughn** [comment]
That command line flag doesn't do at all what I hoped it did. Grr.

This might take a little longer than hoped.

**Pull request closed without merging**